### PR TITLE
Add suppression of type parameters that have notnull constraints

### LINKED ISSUE #462: DiagnosticsSuppress does not suppress CS8634
Given the following constructed example:
```csharp
[Test]
public void Test()
{
    object? possibleNull = GetNext();
    Assert.NotNull(possibleNull);
    DoNothing((object)possibleNull);
}

private static object? GetNext() => new object();

private void DoNothing<T>(T obj)
    where T : class
{}
```
Result in error: 
```
CS8634 The type 'object?' cannot be used as type parameter 'T' in the generic type or method 'TestClass.DoNothing<T>(T)'. Nullability of type argument 'object?' doesn't match 'class' constraint.
```

Even trying to be explicit, doesn't remove the error:
```csharp
    object? possibleNull = GetNext();
    Assert.NotNull(possibleNull);
    object notNull = (object)possibleNull;
    DoNothing(notNull);
```

This might be a limitation on the `DiagnosticSuppressors` where we suppress the error generated for the cast, but we don't change the compiler's understanding of what could be null. We don't do a complete flow analysis on `notNull`

The only option that seems to work is by either using the null suppression operator: `object notNull = possibleNull!;`
or by doing the nullability conversion in a separate method:

```csharp
[Test]
public void Test()
{
    object? possibleNull = GetNext();
    object notNull = EnsureNotNull(possibleNull);
    DoNothing(notNull);
}

private static object? GetNext() => new object();

private static object EnsureNotNull(object? possibleNull)
{
    Assert.NotNull(possibleNull);
    return (object)possibleNull;
}

private void DoNothing<T>(T obj)
    where T : class
{}
```

### PR DESCRIPTION
**Pull request by manfred-brands**
Fixes #462

### CODE REVIEW

File: src/nunit.analyzers/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressor.cs [hunk 1]
```diff
@@ -115,6 +148,23 @@ private static bool ShouldBeSuppressed(SyntaxNode node)
             return false;
         }
 
+        private static bool ShouldBeSuppressed(IMethodSymbol method, string nonMatchedParameterType, InvocationExpressionSyntax invocationExpression)
+        {
+            // if (!Debugger.IsAttached) Debugger.Launch();
```
Comment (mikkelbu) [inline src/nunit.analyzers/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressor.cs:153]:
```suggestion
```

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs [hunk 2]
```diff
@@ -771,5 +772,113 @@ public Lines(string? line1, string? line2)
...
+            ");
+
+            RoslynAssert.Suppressed(suppressor,
+                ExpectedDiagnostic.Create(DereferencePossiblyNullReferenceSuppressor.SuppressionDescriptors[diagnostic]),
+                testCode);
+        }
+
+        [TestCase("class", "CS8634")]
+        [TestCase("notnull", "CS8714")]
+        public void TestIssue462SuppressesMultiple(string constraint, string diagnostic)
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs:831]: I originally had a question about this in the production code, but this test answered it 👍

**Review by mikkelbu — APPROVED** [review]
Only one minor nitpick (the tests answered the rest :) ). I'm happy to merge when you are 👍

File: src/nunit.analyzers/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressor.cs (hunk 1 above)
Comment (manfred-brands) [reply src/nunit.analyzers/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressor.cs]: Removed

**Pull request merged**